cirq-core/cirq/optimizers/merge_interactions.py

```diff
@@ -181,3 +195,45 @@ def _flip_kron_order(mat4x4: np.ndarray) -> np.ndarray:
         for j in range(4):
             result[order[i], order[j]] = mat4x4[i, j]
     return result
+
+
+class MergeInteractions(MergeInteractionsAbc):
+    """Combines series of adjacent one and two-qubit gates operating on a pair
+    of qubits and replaces each series with the minimum number of CZ gates."""
+
+    def __init__(
+        self,
+        tolerance: float = 1e-8,
+        allow_partial_czs: bool = True,
+        post_clean_up: Callable[[Sequence[ops.Operation]], ops.OP_TREE] = lambda op_list: op_list,
+    ) -> None:
+        super().__init__(tolerance=tolerance, post_clean_up=post_clean_up)
+        self.allow_partial_czs = allow_partial_czs
+
+    def _may_keep_old_op(self, old_op: 'cirq.Operation') -> bool:
+        """Returns True if the old two-qubit operation may be left unchanged
+        without decomposition."""
+        if self.allow_partial_czs:
+            return isinstance(old_op.gate, ops.CZPowGate)
+        return isinstance(old_op.gate, ops.CZPowGate) and old_op.gate.exponent == 1
```

Comment:
You just solved another bug in this PR - #4287 :)

However, This results in a change of behavior of `MergeInteractions` optimizer (and hence is potentially a breaking change?)

Let's either make sure not to change the behavior of `MergeInteractions` (and maybe add a comment to the linked issue) or split this PR into two parts?

Reply:
Good catch.  I didn't realize changing to the new way of checking gate type would change behavior.

Reply:
See #4288.  I'd rather fix this before copying the wrong behavior into a new optimizer.

Reply:
Rebased this PR now that the fix is merged.